pkg/ruler/ruler.go

```diff
@@ -741,6 +813,38 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, disabl
 	return result
 }
 
+// filterBackupRuleGroups returns map of rule groups that given instance backs up based on supplied ring.
+// This function only uses User, Namespace, and Name fields of individual RuleGroups.
+//
+// Reason why this function is not a method on Ruler is to make sure we don't accidentally use r.ring,
+// but only ring passed as parameter.
+func filterBackupRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, ring ring.ReadRing, instanceAddr string, log log.Logger, ringCheckErrors prometheus.Counter) []*rulespb.RuleGroupDesc {
```

Comment:
This function seems duplicated, too. Can we generalize the two functions?
Is log the main concern of not reusing the previous function

Reply:
My reasoning for this was that when we implement the Rule evaluation HA that would also result the Rules API HA so we won't be needing `filterBackupRuleGroups` and we can delete it. Deleting an entire unused method is simpler then adjusting existing methods and remove the unused blocks. But thinking now, API HA code and Evaluation HA code can co-exists if we allow cortex to be operated with API HA enabled and Evaluation HA disabled which Evaluation HA is implemented and available. I will make the adjustment.

Reply:
Thanks. It is not a big issue. I am fine to keep it as it is since generalizing the log is probably not easy. But I hope we can generalize `instanceBacksUpRuleGroup`.